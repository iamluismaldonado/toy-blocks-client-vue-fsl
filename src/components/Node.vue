<template>
  <v-expansion-panel class="accordion-node">
    <v-expansion-panel-header>
      <div class="d-flex">
        <div class="d-flex flex-column">
          <span class="accordion-header-name">{{node.name}}</span>
          <span class="accordion-header-url">{{node.url}}</span>
        </div>
        <v-badge class="accordion-badge" left bordered dot inline :color="getColor">
          <span class="text-uppercase accordion-status-text">{{ getStatusText }}</span>
        </v-badge>
      </div>
    </v-expansion-panel-header>
    <v-expansion-panel-content>
      <v-card v-for="block in node.blocks" :key="block.id" class="accordion-expansion rounded-sm" elevation="0" color="#E0DFE0"
              outlined shaped tile>
        <div class="d-flex flex-column ml-2 mt-1 mb-1 mr-2">
          <span class="accordion-expansion-id">{{ block.id }}</span>
          <span class="accordion-expansion-name">{{ block.name }}</span>
        </div>
      </v-card>
    </v-expansion-panel-content>
  </v-expansion-panel>
</template>

<script>
export default {
  name: "node",
  props: {
    node: {
      url: String,
      online: Boolean,
      name: String,
      loading: Boolean,
      blocks: Array,
    },
  },
  data: () => ({}),
  computed: {
    getColor() {
      let badgeColor = "#Eb5757";

      if (this.node.online) {
        badgeColor = "#18cc55";
      }
      return badgeColor;
    },
    getStatusText() {
      let statusText = "Loading";

      if (!this.node.loading) {
        if (this.node.online) {
          statusText = "Online";
        } else {
          statusText = "Offline";
        }
      }
      return statusText;
    },
  },
};
</script>

<style scoped>
.accordion-badge {
  margin-left: auto;
  margin-right: 12px;
  margin-top: 1px;
}
.accordion-header-name {
  font-size: 16px;
  line-height: 24px;
  letter-spacing: 0.44px;
  color: #263238;
}

.accordion-header-url {
  font-size: 14px;
  line-height: 20px;
  display: flex;
  align-items: center;
  letter-spacing: 0.25px;
  color: #263238;
  opacity: 0.54;
}

.accordion-status-text {
  font-weight: 500;
  font-size: 10px;
  line-height: 16px;
  letter-spacing: 1.5px;
  color: #263238;
}
.accordion-expansion {
  margin-top: 5px;
  display: flex;
}

.accordion-expansion-id {
  font-weight: 700;
  font-size: 12px;
  color: #516af8;
}

.accordion-expansion-name {
  font-size: 14px;
}
</style>

<style>
.accordion-node .v-expansion-panels--accordion {
  padding-bottom: 12px;
}

.accordion-node .accordion-badge.v-badge--dot .v-badge__badge {
  height: 6px;
  margin-top: 18px;
  width: 6px;
}

@media only screen and (max-width: 442px) {
  .accordion-node .accordion-badge.v-badge--dot .v-badge__badge {
    margin-top: 28px;
  }
}
</style>