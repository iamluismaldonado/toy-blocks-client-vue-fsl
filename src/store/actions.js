export const actions = {
  async getAllNodes({ commit }, nodeList) {
    for (const el of nodeList) {
      await commit('checkNodeStatusStart', el);

      try {
        const res = await fetch(`${el.url}/api/v1/status`);
        const response = await res.json();
        const params = {
          el,
          name: response.node_name
        };
        await commit('checkNodeStatusSuccess', params);

        try {
          const res = await fetch(`${el.url}/api/v1/blocks`);
          const response = await res.json();
          const params = {
            el,
            blockList: response.data
          };
          await commit('checkNodeBlockSuccess', params);
        }
        catch (e) {
          await commit('checkNodeBlockFailure', el);
        }
      }
      catch (e) {
        await commit('checkNodeStatusFailure', el);
      }
    }
  },
}