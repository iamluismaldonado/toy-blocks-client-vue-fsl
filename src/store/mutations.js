export const mutations = {
  checkNodeStatusStart(state, { url }) {
    let list = state.nodes.list;
    const index = state.nodes.list.findIndex(t => t.url === url);
    if (index >= 0) {
      list = [
        ...state.nodes.list.slice(0, index),
        {
          ...state.nodes.list[index],
          loading: true
        },
        ...state.nodes.list.slice(index + 1)
      ];
    }
    state.nodes = {
      list: [
        ...list,
      ]
    }
  },
  checkNodeStatusSuccess(state, { el: { url }, name }) {
    let list = state.nodes.list;
    const index = state.nodes.list.findIndex(t => t.url === url);
    if (index >= 0) {
      list = [
        ...state.nodes.list.slice(0, index),
        {
          ...state.nodes.list[index],
          name: name,
          online: true,
          loading: false
        },
        ...state.nodes.list.slice(index + 1)
      ];
    }

    state.nodes = {
      list: [
        ...list,
      ]
    }
  },
  checkNodeBlockSuccess(state, { el: { url }, blockList }) {
    let list = state.nodes.list;
    const index = state.nodes.list.findIndex(t => t.url === url);
    if (index >= 0) {
      const blocks = state.nodes.list[index].blocks;
      for (let i = 0; i < blockList.length; i++) {
        blocks.push({ id: blockList[i].id, name: blockList[i].attributes.data });
      }
      list = [
        ...state.nodes.list.slice(0, index),
        {
          ...state.nodes.list[index],
          blocks: blocks
        },
        ...state.nodes.list.slice(index + 1)
      ];
    }

    state.nodes = {
      list: [
        ...list,
      ]
    }
  },
  checkNodeStatusFailure(state, { url }) {
    let list = state.nodes.list;
    const index = state.nodes.list.findIndex(t => t.url === url);
    if (index >= 0) {
      list = [
        ...state.nodes.list.slice(0, index),
        {
          ...state.nodes.list[index],
          online: false,
          loading: false
        },
        ...state.nodes.list.slice(index + 1)
      ];
    }

    state.nodes = {
      list: [
        ...list,
      ]
    }
  },
  checkNodeBlockFailure(state, { url }) {
    let list = state.nodes.list;
    const index = state.nodes.list.findIndex(t => t.url === url);
    if (index >= 0) {
      list = [
        ...state.nodes.list.slice(0, index),
        {
          ...state.nodes.list[index],
          online: false,
          loading: false,
          blocks: []
        },
        ...state.nodes.list.slice(index + 1)
      ];
    }

    state.nodes = {
      list: [
        ...list,
      ]
    }
  },
}