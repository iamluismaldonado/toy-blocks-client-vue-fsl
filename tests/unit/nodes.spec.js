import Vue from 'vue';
import { shallowMount } from '@vue/test-utils'
import Nodes from '@/views/Nodes.vue'
import Node from '@/components/Node.vue'
import store from '../../src/store'
import Vuetify from 'vuetify'

Vue.use(Vuetify)

describe('<Nodes />', () => {
  const defaultProps = {};

  //mocks fetch API
  global.fetch = jest.fn(() =>
    Promise.resolve({
      json: () => Promise.resolve({ el: { url: 'https://thawing-springs-53971.herokuapp.com' }, node_name: 'Test Name' }),
    })
  );

  const render = (props = {}) => shallowMount(Nodes, {
    store,
    propsData: {
      ...defaultProps,
      ...props
    }
  });

  it('renders', () => {
    const wrapper = render();
    expect(wrapper.find('.blocks-wrapper').exists()).toBeTruthy;
  });

  it('match the snapshot', () => {
    const wrapper = render();
    expect(wrapper).toMatchSnapshot();
  });

  it('renders nodes', () => {
    const wrapper = render();
    expect(wrapper.findComponent(Node).exists()).toBeTruthy;
  });
});