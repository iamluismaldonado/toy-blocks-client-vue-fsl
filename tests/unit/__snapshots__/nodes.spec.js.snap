// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Nodes /> match the snapshot 1`] = `
<div class="d-flex flex-column blocks-wrapper">
  <h1 class="main-title">Nodes</h1>
  <v-row-stub tag="div" justify="center" class="accordion">
    <v-expansion-panels-stub activeclass="v-item--active" accordion="true">
      <node-stub node="[object Object]"></node-stub>
    </v-expansion-panels-stub>
    <v-expansion-panels-stub activeclass="v-item--active" accordion="true">
      <node-stub node="[object Object]"></node-stub>
    </v-expansion-panels-stub>
    <v-expansion-panels-stub activeclass="v-item--active" accordion="true">
      <node-stub node="[object Object]"></node-stub>
    </v-expansion-panels-stub>
    <v-expansion-panels-stub activeclass="v-item--active" accordion="true">
      <node-stub node="[object Object]"></node-stub>
    </v-expansion-panels-stub>
  </v-row-stub>
</div>
`;
